### PR TITLE
feat: New Kura Networking

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

This PR starts the implementation of a new version of the Kura Networking based on NetworkManager.
In particular:

- the `org.eclipse.kura.nm` bundle is added; it contains the `NMConfigurationServiceImpl` that is responsible to apply the configuration to the system. The calls to NM are still missing and they will be implemented in a separate PR.
- the bundle contains a writer and monitor for the DHCP server, since NM doesn't provide a full-featured server.
- the `org.eclipse.kura.net.configuration` bundle is added to provide common code shared between old and new `NetworkConfigurationService`
- the `net.admin` is maintained in the code, since some devices are not equipped with NM and they will continue working with the old Kura Networking. 

Currently the default configuration is to keep the old networking. To enable the NM based one, build the distrib project wih the `-Dkura.network.provider=network-manager` option. Morever, NM has to be manually started after the installation in the target device.

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

### CODE REVIEW

File: kura/distrib/src/main/resources/raspberry-pi-arm64/kura_install.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh
 #
-# Copyright (c) 2022 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/distrib/src/main/resources/raspberry-pi-arm64/kura_install.sh]: no need to change this probably

File: kura/org.eclipse.kura.linux.net/src/main/java/org/eclipse/kura/linux/net/dhcp/DhcpServerManager.java [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /*******************************************************************************
  * Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
- *
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.linux.net/src/main/java/org/eclipse/kura/linux/net/dhcp/DhcpServerManager.java]: no need to change this file

File: kura/org.eclipse.kura.net.admin.firewall/src/main/java/org/eclipse/kura/net/admin/FirewallConfigurationService.java [hunk 3]
```diff
@@ -1,5 +1,9 @@
 /*******************************************************************************
+<<<<<<< HEAD
  * Copyright (c) 2016, 2023 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.net.admin.firewall/src/main/java/org/eclipse/kura/net/admin/FirewallConfigurationService.java]: merge issues?

File: kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml [hunk 4]
```diff
@@ -23,5 +23,5 @@
       <provide interface="org.eclipse.kura.security.FloodingProtectionConfigurationService"/>
    </service>
    <property name="kura.ui.service.hide" type="Boolean" value="true"/>
-   <reference bind="setFirewallConfigurationService" unbind="unsetFirewallConfigurationService" cardinality="1..1" interface="org.eclipse.kura.net.admin.FirewallConfigurationService" name="FirewallConfigurationService" policy="static"/> 
+   <reference bind="setFirewallConfigurationService" unbind="unsetFirewallConfigurationService" cardinality="1..1" interface="org.eclipse.kura.net.admin.FirewallConfigurationService" name="FirewallConfigurationService" policy="static"/>
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml:26]: is there any non formatting-related change to this file?

File: kura/org.eclipse.kura.nm/OSGI-INF/networkConfigurationService.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+   Copyright (c) 2011, 2022 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.nm/OSGI-INF/networkConfigurationService.xml]: please update copyright header

File: kura/org.eclipse.kura.nm/build.properties [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#
+#  Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.nm/build.properties]: please update copyright header

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/event/NetworkConfigurationChangeEvent.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/configuration/event/NetworkConfigurationChangeEvent.java]: please update copyright header

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/GwtNetInterfaceConfigBuilder.java [hunk 8]
```diff
@@ -132,7 +132,6 @@ private void setWifiProperties() {
             if (wifiMode.equals(GwtWifiWirelessMode.netWifiWirelessModeAccessPoint.name())) {
                 setWifiMasterProperties();
             }
-
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/GwtNetInterfaceConfigBuilder.java]: no need to change this?

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/NetworkConfigurationServiceProperties.java [hunk 9]
```diff
@@ -108,7 +108,6 @@ public String getIp4Netmask(String ifname) {
                 return NetworkUtil.getNetmaskStringForm(prefix.intValue());
             }
         }
-
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/NetworkConfigurationServiceProperties.java]: no need to change this?

File: kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/NetworkConfigurationServicePropertiesBuilder.java [hunk 10]
```diff
@@ -129,9 +129,9 @@ private void setWifiMasterProperties() {
         this.properties.setWifiMasterGroupCiphers(this.ifname, Optional.ofNullable(gwtWifiConfig.getGroupCiphers()));
 
         // wifi master specific properties
-
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/net2/configuration/NetworkConfigurationServicePropertiesBuilder.java:132]: no need to change this?

File: kura/test/org.eclipse.kura.net.admin.test/src/test/java/org/eclipse/kura/net/admin/NetworkConfigurationServiceImplTest.java [hunk 11]
```diff
@@ -87,6 +87,9 @@ protected void initVisitors() {
         ComponentContext componentContextMock = mock(ComponentContext.class);
         BundleContext bundleCtxMock = mock(BundleContext.class);
         when(componentContextMock.getBundleContext()).thenReturn(bundleCtxMock);
+        UsbService usbService = mock(UsbService.class);
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.net.admin.test/src/test/java/org/eclipse/kura/net/admin/NetworkConfigurationServiceImplTest.java:90]: please update copyright header

File: kura/test/org.eclipse.kura.nm.test/build.properties [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#
+# Copyright (c) 2017, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.nm.test/build.properties]: please update copyright header

File: kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml (hunk 4 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml:26]: No, probably the auto-formatter has removed the trailing space.

File: kura/org.eclipse.kura.net.configuration/build.properties [hunk 13]
```diff
@@ -0,0 +1,22 @@
+#
+#  Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.net.configuration/build.properties]: Please update copyright header

File: kura/org.eclipse.kura.net.configuration/src/main/resources/org/eclipse/kura/net/configuration/messages/NetworkConfigurationMessagesBundle.properties [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#
+#  Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.net.configuration/src/main/resources/org/eclipse/kura/net/configuration/messages/NetworkConfigurationMessagesBundle.properties]: Please update copyright header

File: kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml (hunk 4 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.network.threat.manager/OSGI-INF/flooding_configuration_service.xml:26]: Then can you please either revert the changes to this file or update the copyright header?

File: kura/test/org.eclipse.kura.net.configuration.test/.java-version [hunk 15]
```diff
@@ -0,0 +1 @@
+1.8
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.net.configuration.test/.java-version]: What is the purpose of this file?

File: kura/test/org.eclipse.kura.nm.test/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 16]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.nm.test/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: What is the purpose of this file?

File: kura/test/org.eclipse.kura.net.configuration.test/.java-version (hunk 15 above)
Comment (pierantoniomerlino) [reply kura/test/org.eclipse.kura.net.configuration.test/.java-version]: This a spourius file. I'll update the .gitignore file to ignore it.

File: kura/test/org.eclipse.kura.nm.test/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 16 above)
Comment (pierantoniomerlino) [reply kura/test/org.eclipse.kura.nm.test/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: This is needed to mock static classes, as presented [here](https://www.baeldung.com/mockito-mock-static-methods).

**Review by nicolatimeus — APPROVED** [review]

**Comment by MMaiero** [comment]
Waiting for a successful build before merging.
The duplication reported by Sonar is accepted by the team due to the fact that the firewall code needed to be duplicated in two separate bundles to keep consistency and not make breaking changes.

**Pull request merged**